### ISSUE TITLE
notebook() does not automatically launch/open a browser

### ISSUE THREAD

**Issue by mgiugliano** [JuliaLang/IJulia.jl#551]
Instead of leaving the environment vars PYTHON and JUPYTER empty when installing PyCall and IJulia (as suggested when no Python/Jupyter install is available),
```
ENV["PYTHON"] = ""
ENV["JUPYTER"] = ""
```
I have specified the full path to my freshly installed Python 3.6 and Jupyter 4.2.1 (by Anaconda 4.3.1). By definition, this does not install the miniconda Python (2) private environment.

Unfortunately, although Jupyter seems to work fine with the Julia's kernel, the commands
```
using IJulia
notebook()
``` 
no longer automatically open/launch the web browser on my Mac. That was a default behaviour by the miniconda (automated private) install. This forces me to manually navigate to http://localhost:8888/ or type at the os command line:  `jupyter notebook list`

macOS (x86_64-apple-darwin13.4.0)
Julia Version 0.5.2
IJulia                        1.5.0
PyCall                      1.12.0
Python 3.6
Jupyter 4.2.1

It might be related to http://www.andrewjaffe.net/blog/2017/05/python-bug-hunt.html and https://github.com/jupyter/notebook/issues/2438, although with the miniconda install it was working. 

I temporarily solved it by typing at a shell prompt:
`> jupyter notebook --generate-config`
and then uncomment/edit the line of the file ~/.jupyter/jupyter_notebook_config.py (just created), into
`c.NotebookApp.browser = 'Safari'` (to have Safari explicitly indicated)

**Comment by stevengj** [comment]
IJulia.notebook_cmd will tell you the command-line arguments it is using in `notebook()`; does running this manually work for you?

I suspect that this is a Jupyter bug, though I'm not sure why it would work at the command line.

**Comment by stevengj** [comment]
By the way, it doesn't matter what environment variables you set when you *call* IJulia, only when you *build* IJulia.   Also, the PYTHON environment variable doesn't matter.

Try
```
ENV["JUPYTER"]="jupyter"
Pkg.build("IJulia")
```

**Comment by mgiugliano** [comment]
Indeed, it might not be a IJulia bug, but IJulia is affected.

I am 90% sure that leaving ENV["JUPYTER"]="" initiate a download of miniconda. :-(
I don't want miniconda.. ;-) I don't want extra space used in .julia/

**Comment by stevengj** [comment]
`ENV["JUPYTER"]=""` before `Pkg.build` will indeed case it to use Conda.  But I didn't suggest `ENV["JUPYTER"]=""`, I suggested `ENV["JUPYTER"]="jupyter"` (assuming `jupyter` is in your PATH, as your above snippets implied).

Again, I want to verify that you set the JUPYTER env var before `Pkg.build`, *not* before `using IJulia`.   The environment variable *only* affects the `Pkg.build`.

**Comment by rsrock** [comment]
I have just run into this issue on two different machines. I suspect the problem is upstream, as I also have the problem running jupyter from the shell.

First, on a student's machine with a brand-new install of the macOS 5.2 binary, new package dir, and using the IJulia default miniconda install.  On her machine, `IJulia.notebook_cmd` pointed at the Conda install in the package dir. Manually navigating to `http://localhost:8888` brought up the Jupyter notebook home page.

I also just checked using `notebook()` from my julia-0.6-rc2, and the same thing.  The server runs, but doesn't load the page in the browser automatically.  This has worked in the past, but I usually launch jupyter from the shell instead so I'm not sure when this changed.  I'm using my own Anaconda jupyter, and Pkg.build has picked up the correct location:

```
julia> IJulia.notebook_cmd
2-element Array{String,1}:
 "/Users/rrock/anaconda/bin/jupyter"
 "notebook"
```

I just checked running jupyter from the shell, and I see the same problem there too (so the problem is probably upstream).  I see the following error when running from the shell:

```
0:97: execution error: "http://localhost:8888/tree?token=xxxxxxxxxxxxxxxxxxxx" doesn’t understand the “open location” message. (-1708)
```

**Comment by rsrock** [comment]
Apparently, the problem is with macOS 10.12.5, and will be fixed in 10.12.6.  A temporary workaround is to run

```
export BROWSER=open
```

in the shell before launching julia.

**Comment by stevengj** [comment]
@rsrock, it might be reasonable to include that BROWSER workaround in `IJulia.notebook()`.  Change the line `p = spawn(Cmd(`$notebook_cmd`, detach=true, dir=dir))` to
```jl
if is_apple() # issue #551 workaround
    withenv("BROWSER"=>"open") do
        p = spawn(Cmd(`$notebook_cmd`, detach=true, dir=dir))
    end
else
    p = spawn(Cmd(`$notebook_cmd`, detach=true, dir=dir))
end
```

**Comment by rsrock** [comment]
Yes, that works here.

**Comment by stevengj** [comment]
A PR would be welcome.

**Comment by stevengj** [comment]
Unless we should just wait for 10.12.6?  I don't see any likely downside to including the workaround, however.

**Comment by Ryanglambert** [comment]
This is happening for python 2.7.13 jupyter version 4.3.0.  

Any other information I could provide?

**Comment by stevengj** [comment]
@Ryanglambert see above: assuming you are on MacOS, it is a MacOS bug.  You can either ignore it (just open the localhost:8888 url in your browser) or set `ENV["BROWSER"]="open"` before running `IJulia.notebook()`.

**Comment by stevengj** [comment]
Closed by #552.

**Comment by harald-gegenfurtner** [comment]
I can confirm that this problem occured right after the update to macOS 10.12.5.

My workaround is to define an alias 'nb' in ~/.bash_profile like this:

`alias nb='export BROWSER=open; jupyter notebook'`

**Comment by typhoonbxq** [comment]
I found one solution:

Step 1 generate the config file
$ jupyter notebook --generate-config

Step 2 open the config file
$ sudo vi ~/.jupyter/jupyter_notebook_config.py

Step3 find c.NotebookApp.browser set it to  u'Safari', also remember to remove '#' at this line 
#c.NotebookApp.browser = u' '   --->  c.NotebookApp.browser = u'Safari'

**Comment by stevengj** [comment]
The workaround in #552 will only work for running `IJulia.notebook()` from within Julia; if you are running `jupyter` from the command line you will still need a manual workaround.

**Comment by harald-gegenfurtner** [comment]
As far as I can tell, the bug has already been fixed with the macOS 10.12.6 update.
`using IJulia; notebook()` works fine. Thank you!
